### PR TITLE
ENH Add "adversarial debiasing" 2nd PR

### PR DESCRIPTION
**Pull request by SeanMcCarren**
A clean slate PR of #973 so feedback is cleared. 
- Base class is made private.
- Subclass API is more digestible as we removed arbitrary choices of loss functions and preprocessors from this description.

### CODE REVIEW

**Comment by adrinjalali** [comment]
@SeanMcCarren could you please merge the latest `main` and run `black` on your code? (you can also enable git pre-commit hooks since now we have them)

@MiroDudik when do you think you'd have a chance to look at this. It'd be nice to merge this one.

**Comment by SeanMcCarren** [comment]
@adrinjalali apologies I missed this comment! Done. Let's see what the CI does!

**Comment by adrinjalali** [comment]
That's this error I think: https://github.com/scikit-learn/scikit-learn/pull/22529

It's fixed upstream, and we should be doing a release this week.

**Comment by adrinjalali** [comment]
FYI, the fix on the sklearn side is released.

**Comment by adrinjalali** [comment]
FYI, merging the latest `main` should fix the issue of building the docs so that this PRs changes are visible on circleci's artifacts.

File: fairlearn/adversarial/_adversarial_mitigation.py
Comment (MiroDudik) [inline fairlearn/adversarial/_adversarial_mitigation.py:887]: I think we should remove this parameter (see [this comment](https://github.com/fairlearn/fairlearn/pull/973#discussion_r851659798)). Since we always default to either binary or multinomial logistic loss, the predictor function is always determined from those.

File: fairlearn/adversarial/_adversarial_mitigation.py
Comment (MiroDudik) [inline fairlearn/adversarial/_adversarial_mitigation.py:900]: Once we remove `predictor_function`, we can also remove this one.

File: docs/user_guide/adversarial.rst
Comment (MiroDudik) [inline docs/user_guide/adversarial.rst:44]: In the current implementation, the user doesn't choose loss functions, so you could either remove the section, or possibly describe the logic that is used to determine the loss function from the data type (e.g., discrete data will result either in binary or multinomial logistic loss, and continuous data into square loss).

**Review by MiroDudik — CHANGES_REQUESTED** [review]
My review just focused on the API and documentation. I the PR is really close. My only substantive comment is to remove `predictor_loss` and `threshold_value` from the classification class. Just like with the other customizable components, the idea is to hide that functionality in the current version.

One possible idea to retain `threshold_value` functionality would be to implement `predict_proba`, but that can wait for a future release.

**Comment by SeanMcCarren** [comment]
@MiroDudik Thanks again for the round of feedback :)

I'll get back to it starting the 28th

File: fairlearn/adversarial/_adversarial_mitigation.py
Comment (SeanMcCarren) [reply fairlearn/adversarial/_adversarial_mitigation.py:846]: @MiroDudik Wait, you mean `predictor_loss` instead of `predictor_function` right?

File: fairlearn/adversarial/_adversarial_mitigation.py
Comment (SeanMcCarren) [reply fairlearn/adversarial/_adversarial_mitigation.py:900]: @MiroDudik You mean `predictor_loss` ?

**Comment by SeanMcCarren** [comment]
@MiroDudik Hey Miro, I lost track of our discussion and I want to make sure I understand your feedback correctly. Could you quickly mention whether you want to remove `predictor_loss`, `predictor_function`, or both, as I am not sure which one to remove. Thanks! If I remember correctly your point was to remove the `predictor_loss` and `predictor_function` for the classifier, but **not** for the regressor, correct? Also, we aren't touching the `adversary_loss`, correct?

File: fairlearn/adversarial/_adversarial_mitigation.py
Comment (MiroDudik) [inline fairlearn/adversarial/_adversarial_mitigation.py:1055]: This should be dropped.

File: fairlearn/adversarial/_adversarial_mitigation.py
Comment (MiroDudik) [reply fairlearn/adversarial/_adversarial_mitigation.py:846]: Sorry for the confusion... I actually did mean removing `predictor_function`.

**Comment by MiroDudik** [comment]
> @MiroDudik Hey Miro, I lost track of our discussion and I want to make sure I understand your feedback correctly. Could you quickly mention whether you want to remove `predictor_loss`, `predictor_function`, or both, as I am not sure which one to remove. Thanks! If I remember correctly your point was to remove the `predictor_loss` and `predictor_function` for the classifier, but **not** for the regressor, correct? Also, we aren't touching the `adversary_loss`, correct?

Sorry for the delay on this. I'm still in agreement with my [original comment](https://github.com/fairlearn/fairlearn/pull/973#discussion_r851659798). To confirm: my suggestion was to remove `predictor_function` (and therefore also `threshold_value`), and also remove `predictor_loss` and `adversary_loss` in both classification and regression; this means that the losses would be restricted to logistic loss (for binary or categorical data) or square loss (for continuous data, univariate or multivariate); the loss would be automatically inferred from the data type.

But you remember correctly that although I'm in favor of dropping `predictor_loss` even in regression, I don't feel too strongly about it and if you'd like to allow that flexibility, please go for it.

I don't want to allow `adversary_loss` in any of these, because the standard use case for it is that sensitive features are categorical and so they need to get transformed etc., which would make the specification of the loss a bit confusing.

**Comment by hildeweerts** [comment]
Copying @riedgar-ms's comment in Discord in case it's helpful: "The doctest failure is due to pandas deciding to change its defaults for printing out DataFrames. Look for pd.set_option in the Intersecting Groups page for the fix"

**Comment by hildeweerts** [comment]
If I recall correctly @SeanMcCarren ran into an issue with the tensorflow tests before - is this still the same one or a new error?

**Comment by SeanMcCarren** [comment]
@hildeweerts I don't recognize it, so I think something this is a new problem? But I don't know why exp_grad test is breaking.

**Comment by riedgar-ms** [comment]
@SeanMcCarren have you at least been able to straighten out that pandas display issue?

**Comment by riedgar-ms** [comment]
@SeanMcCarren if once you're resynced to `main` the only remaining issue is with `test_othermlpackages` we will probably go ahead and merge. On closer inspection, that build has been failing since July, and we need to migrate it to GitHub Workflows anyway.

Tagging @fairlearn/fairlearn-maintainers (especially @adrinjalali ) for info.

**Comment by SeanMcCarren** [comment]
Thanks for the fix @riedgar-ms !

File: fairlearn/adversarial/_adversarial_mitigation.py
Comment (SeanMcCarren) [reply fairlearn/adversarial/_adversarial_mitigation.py:812]: Ok, `predictor_function` or `threshold_value` are no longer visible or changeable in the public API, but I did not remove them from the base class.

**Comment by SeanMcCarren** [comment]
@MiroDudik @riedgar-ms @hildeweerts 

Except for the codecov/patch and codecov/project failures (which I think are too weird to avoid for this PR), there is a build-test failing for some weird reason (seems to be codecov related as well 🤔). Any idea what is going on here?

Also, if there is anything I can do for this PR, I have much more time this week than last! :) So any last suggestions or wishes? Otherwise, I would say this is finished (for now?)

**Comment by adrinjalali** [comment]
The build test was failing due to upload issue to codecov, it happens sometimes. I restarted the job.

**Comment by riedgar-ms** [comment]
Just pulled in my update to the codecov build. See if that makes things better

**Comment by SeanMcCarren** [comment]
Yeah it seems to make it a bit better @riedgar-ms :)

**Comment by SeanMcCarren** [comment]
Should I squash commits? :)

**Comment by adrinjalali** [comment]
We squash and merge anyway @SeanMcCarren , no need for you to squash and force push here :)

**Comment by riedgar-ms** [comment]
The code coverage looks like there are still a couple of 'functional' (rather than 'error') branches which aren't tested. However, if the general consensus is to merge now, I don't object @adrinjalali @MiroDudik

**Comment by adrinjalali** [comment]
I'm also happy for this to be merged and see what feedback we get.

**Comment by riedgar-ms** [comment]
@MiroDudik are you OK with this being merged now?

**Comment by SeanMcCarren** [comment]
> I'm also happy for this to be merged and see what feedback we get.

Yes me too haha. Especially because the API is still on the heavy side

File: fairlearn/adversarial/_adversarial_mitigation.py
Comment (MiroDudik) [inline fairlearn/adversarial/_adversarial_mitigation.py:1086]: The constructor should match the documentation above.

File: fairlearn/adversarial/_adversarial_mitigation.py
Comment (MiroDudik) [inline fairlearn/adversarial/_adversarial_mitigation.py:821]:
```suggestion
    The predictor model takes the features :code:`X` as input and seeks
    to predict :code:`y`.
    For binary classification, the predictor model should return a single
    real-valued score, which is transformed into a probability of the
    positive class via the logistic function (aka sigmoid), similarly to
    logistic regression. For multi-class classification, the predictor
    model should return a vector of real-valued scores, which are
    transformed into class probabilities via the softmax function,
    similarly to multinomial logistic regression. The training loss is
    the negative log likelihood (aka log loss, logistic loss,
    cross-entropy loss).
    
    The adversarial model for demographic parity takes scores
    produced by the predictor model as input, and seeks to predict
    :code:`sensitive_features`. Depending on the type of the provided
    sensitive features, the model should produce a scalar
    or vector output. Three types of sensitive features are supported:
    (1) a single binary feature; (2) a single discrete feature; (3) one or
    multiple real-valued features. For a single binary sensitive feature
    and a single discrete feature, the network outputs are transformed
    by the logistic function and the softmax function, respectively, and
    the loss is the negative log likelihood. For one or multiple
    real-valued features, the network output is left as is, and the
    loss is a square loss.
```

File: fairlearn/adversarial/_adversarial_mitigation.py
Comment (MiroDudik) [inline fairlearn/adversarial/_adversarial_mitigation.py:960]: Please adjust the description below similarly to the description I'm suggesting for the classifier above. The suggestions for the parameter descriptions below are exactly the same as for `AdversarialFairnessClassifier`.

File: fairlearn/adversarial/_adversarial_mitigation.py
Comment (MiroDudik) [inline fairlearn/adversarial/_adversarial_mitigation.py:78]: Please adjust the parameter descriptions similar to `AdversarialFairnessClassifier`.

File: fairlearn/adversarial/_adversarial_mitigation.py
Comment (SeanMcCarren) [reply fairlearn/adversarial/_adversarial_mitigation.py:946]: Ah okay

**Comment by riedgar-ms** [comment]
@SeanMcCarren you appear to have a couple of test failures

**Review by MiroDudik — COMMENTED** [review]
I have made small tweaks in the API documentation. @SeanMcCarren : please check especially that I got things right for how the equalized odds is implemented (there's a new paragraph at the end of the constructor description in each of the three AdversarialFairness classes).

So now I'm happy with the API and API documentation.

I still need to make a pass through user_guide/adversarial.rst, but if that's the only item left, then I think that it's fine to merge (the user guide can be updated also once the new package is released).

**Review by MiroDudik — APPROVED** [review]

**Comment by riedgar-ms** [comment]
@SeanMcCarren can you check on those two failing tests?

**Comment by SeanMcCarren** [comment]
Yes, I will!

**Comment by SeanMcCarren** [comment]
@riedgar-ms Oops small mistake. But somehow codecov is failing again now haha.

**Comment by riedgar-ms** [comment]
If it's only a small decrease, I'm prepared to overlook it

**Comment by SeanMcCarren** [comment]
If you have any suggestions to improve the test coverage, I can take a look (possible after merge as well considering the hurry). I had a lot of trouble because so much of the code requires pytorch or tensorflow.

**Comment by riedgar-ms** [comment]
You should be able to put the tests in the `test_othermlpackages` directory.

Looking on codecov, in `_pytorch_engine.py`, it appears that `shuffle()`, `evaluate()` and `trainstep()` are essentially untested. We might not be able to test the CUDA branches, but I think we should have something for the above.

Although it would be worth seeing how my PR to move the otherml workflow affects things.

**Comment by riedgar-ms** [comment]
OK, now that the `other-ml` build is running and uploading results, the code coverage doesn't look quite so bad. A lot of the missing lines are the exception cases, or the CUDA handling. There are a few bits which still look 'functional' without coverage, but at this point, I'd be willing to see those have a separate PR. @adrinjalali ? @hildeweerts ? @MiroDudik ?

**Comment by riedgar-ms** [comment]
@MiroDudik are you done with the tweaks you wanted? Is there anything else from your PoV?

**Comment by MiroDudik** [comment]
Almost done. I'd like to do a final pass Wednesday 10/26 morning.

**Comment by riedgar-ms** [comment]
@MiroDudik we can follow up in the community call on Thursday then :-)

**Comment by MiroDudik** [comment]
I'm done with editing. I'm good to get this merged in the current form.

**Review by hildeweerts — APPROVED** [review]

**Comment by riedgar-ms** [comment]
Since you're making updates @MiroDudik , do you want to add a note to the v0.8 changes:
https://github.com/fairlearn/fairlearn/blob/main/docs/user_guide/installation_and_version_guide/v0.8.0.rst

**Comment by MiroDudik** [comment]
@SeanMcCarren : can you check that I have described the transformation, loss function, and prediction correctly in the table here:
https://output.circle-artifacts.com/output/job/31be9098-ea24-4c31-8fcc-83c3e4752c0c/artifacts/0/docs/_build/html/user_guide/mitigation.html#data-types-and-loss-functions

(I don't think that we need to list the internal keywords)

**Comment by MiroDudik** [comment]
@SeanMcCarren confirmed that the table is correct, so once the tests pass, please merge!

**Pull request merged**
